compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirPrimaryConstructorSuperTypeChecker.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.KtNodeTypes
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.toRegularClass
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getParent
+import org.jetbrains.kotlin.fir.declarations.FirConstructor
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.isInterface
+import org.jetbrains.kotlin.fir.declarations.primaryConstructor
+import org.jetbrains.kotlin.fir.types.coneType
+import org.jetbrains.kotlin.fir.types.impl.FirImplicitAnyTypeRef
+import org.jetbrains.kotlin.utils.addToStdlib.lastIsInstanceOrNull
+
+object FirPrimaryConstructorSuperTypeChecker : FirRegularClassChecker() {
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.isInterface) {
+            return
+        }
+
+        val primaryConstructor = declaration.primaryConstructor
+
+        if (primaryConstructor == null) {
+            checkSupertypeInitializedWithoutPrimaryConstructor(declaration, reporter, context)
+        } else {
+            checkSuperTypeNotInitialized(primaryConstructor, declaration, context, reporter)
+        }
+    }
+
+    private fun checkSuperTypeNotInitialized(
+        primaryConstructor: FirConstructor,
+        regularClass: FirRegularClass,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        val containingClass = context.containingDeclarations.lastIsInstanceOrNull<FirRegularClass>()
+        val delegatedConstructorCall = primaryConstructor.delegatedConstructor ?: return
+        // No need to check implicit call to the constructor of `kotlin.Any`.
+        if (delegatedConstructorCall.constructedTypeRef is FirImplicitAnyTypeRef) return
+        val superClass = delegatedConstructorCall.constructedTypeRef.coneType.toRegularClass(context.session) ?: return
+        // Subclassing a singleton should be reported as SINGLETON_IN_SUPERTYPE
+        if (superClass.classKind.isSingleton) return
+        if (regularClass.isEffectivelyExpect(containingClass, context) ||
+            regularClass.isEffectivelyExternal(containingClass, context)
+        ) {
+            return
+        }
+        if (delegatedConstructorCall.source?.elementType != KtNodeTypes.SUPER_TYPE_CALL_ENTRY) {
```

Comment:
Is it possible to instead check `delegatedConstructorCall.source?.kind is FirRealPsiSourceElement`?

Reply:
We can and it will work.

But I think this check is slightly better because it's closer to what we actually want. Ideally, I guess we should add a boolean property `isSuperTypeCall` to the `FirDelegatedConstructorCall` and set the value when builiding the FIR element. But it seems the current check on the type is sufficient.

On the other hand, in the caes of a `SUPER_TYPE_ENTRY`, I really think the delegatedConstructorCall should have referenced this `SUPER_TYPE_ENTRY` instead. Ideally, it shouldn't be a fake source in that case. But fixing that is not trivial because you need resolution to happen in order to tell a class from an interface and then you need to replace source. It looks like nowhere else will ever need to read that source so I didn't make this change.

Therefore, I think relying on the source being fake is not as clear.